### ISSUE TITLE
[REQUEST] delay effect additions

### ISSUE THREAD

**Issue by altruios** [sadko4u/lsp-plugins#136]
Bread-crumbs thanks to:
https://github.com/wwmm/pulseeffects/issues/753

I was hoping to ask you to add some features to this plug-in.

can you increase the maximum delay length from 1000ms to something significantly higher like 99,999ms.

in addition, though it might be a bit of work, can there be a 'tap' feature, for the delay and other plug ins... where a 'tap' is a tab of said effect in that tab window, allowing for multiple delays at multiple individual settings. 

the first change should be easy... hopefully?
and the second is a cool nifty thing to have, but also sounds like a fair bit of work.

basically I want to use the plugin to do this song live on discord:
queen: the prophet's song... link below.
https://youtu.be/HzdjMLKKdgk?t=209

and currently the maximum value of delay is too short for this song. needs to be able to go up to 3350ms at minimum :)

**Comment by sadko4u** [comment]
@altruios 

The compensation delay was developed for a bit another purposes: to adjust phase difference between two tracks. For example, you have drum multi-mic recording. To make drums sounding well, you need adjust the delay for each track to make tracks in-phase. That's why the delay allows to set time offset in distance units (meters and centimeters), samples and milliseconds. Moreover, delay of 100 seconds requires huge memory buffers to be allocated (about 20 MB at 48 kHz sample rate).

Allocating such buffer for phase correction purposes is too generously.

I think better would be to implement another one delay plugin which will fit your requirements instead of modifying **comp_delay_x2** plugin.

FYI, there's also another plugin which works with short delays and allows to make some spatial effects.
https://lsp-plug.in/?page=manuals&section=slap_delay_stereo

**Comment by altruios** [comment]
ah, copy. did not consider memory.
I'm not familar with what a good target number would be.
but would a 10-second limit at ~2MB be more reasonable, or is that still too wild?

the phase difference two tracks makes sense though - it was a hope.

**Comment by sadko4u** [comment]
> but would a 10-second limit at ~2MB be more reasonable, or is that still too wild?

I think yes. Since we require many phase corrections when mixing multi-miced tracks, it will be a huge use of memory. Moreover, adding TAP button for this plugin is something unnatural.

I'm thinking more on implementing another one plugin which will allow to do the stuff you're asked to do.

**Comment by sadko4u** [comment]
@altruios 
I think we can start gathering information about functions that new delay should provide. Can you give a list of them with detailed explanation of each function?

**Comment by altruios** [comment]
aye -I have something written up. it is not super detailed. if this is not what you want, give me a format to follow :)


delay:
"TAP" pattern:
    ~each TAP should have the following parameters:
        "wet/dry" signal sliders.
        "delayTime" in ms. === maximum should be at least 3350.
        "delayTimeDetermined" is a bool that overwrites delaytime with the following formula
            index Of Tap * tap[0].delayTime //resulting in delay that is "on beat"
        "stereoPan": position center of signal.
        "bypass": individual TAP bypass toggle

    ~add and remove TAPS as needed.
    ~generally TAPS are displayed as tabbed windows, with the tab being on the top.
"bypass":

this is just my initial thoughts, not a complete list or a reasonable blueprint.

~thoughts on efficency or squeezing out bytes
    note I am an idiot.
        can the TAPS each read from the same stream/audio source?
            if that's possible limits are determined by delayTime instead of the count of current TAPS or something more complicated.

**Comment by sadko4u** [comment]
Do I right that you need multiple delay lines where you can set up delay for each delay line by tapping it several times?
Is it reasonable to have Wet/Dry slider for each delay line? Because summing all these delay lines with 0 dB Dry level will give very loud sound.
"delayTimeDetermined" - I believe the better way would be to provide just a multiplier and not use an index as multiplier. I also can add a reference: which delay line to use as a tap source.

Also, since plugin has a limited number of controls, what's the possible maximum number of Taps? Is 16 enough?

**Comment by altruios** [comment]
Aye 16 is plenty I think.
   I’ll think more on the dry/wet. That’s a good point l...There’s a good answer somewhere in my head.

The determined delay idea I haven’t seen before - just something I’ve wanted when working with delays. And thinking more on it - it would work better as a master toggle, instead of individual tap. The idea of it is to set the delay once (on the first tap) and all subsequent taps are n*TAPS[0].delayTime where n starts at 2, and n++ for each tap. So a 1650ms delay with 3 taps total would result in 1625ms, 3350ms, and 4875ms. When it is toggled on - only the first delay should be editable - and other values get calculated after changing the first. 
(Sorry for the rant - weather you base it on the index(actually index+1)  or a multiplier is up to you)

Tap tempo is something that should be added (forgot about that as I hardly ever use it) and yeah - that should be on a per tap level.

**Comment by sadko4u** [comment]
I've got the delay prototype plugin:
![image](https://user-images.githubusercontent.com/12834464/101843294-f5b04f00-3b5a-11eb-855e-566d5dee837c.png)

- Allows to select up to 8 different tempos.
- Allows to use up to 16 delay lines.
- Each delay line allows to compute the delay time based on:
  - another delay processor's time;
  - specified tempo and notation-based delay;
  - exact delay time specified in seconds;
- Each delay line allows to set the feedback delay which should be not greater than the processor's self delay.
- Up to 256 seconds delay is possible. To not to consume a lot of memory, the maximum delay time is configured via dedicated combo.
- Gain, panning and delay values are smoothed and can be automated without clicks/pops/etc.

Any ideas of what else can be added to the plugin?

**Comment by altruios** [comment]
This looks amazing.

As far as ideas... only crazy ones that would (probably) never be used.

This is fantastic!

**Comment by sadko4u** [comment]
Merged all related changes into **devel** branch: 819a54b2bef5e7453d424519a6fdac74dba13cb2

**Comment by sadko4u** [comment]
Available in 1.1.28 release!
